Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -957,16 +959,31 @@ class BrowserViewController: UIViewController {
         }
     }
     // Mark: Opening New Tabs
+    
+    var isInPrivateMode: Bool {
+        return tabTrayController?.privateMode ?? tabManager.selectedTab?.isPrivate ?? false
+    }
 
     @available(iOS 9, *)
-    func switchToPrivacyMode(isPrivate isPrivate: Bool ){
-        applyTheme(isPrivate ? Theme.PrivateMode : Theme.NormalMode)
-
-        let tabTrayController = self.tabTrayController ?? TabTrayController(tabManager: tabManager, profile: profile, tabTrayDelegate: self)
-        if tabTrayController.privateMode != isPrivate {
-            tabTrayController.changePrivacyMode(isPrivate)
+    func switchToPrivacyMode(isPrivate isPrivate: Bool) -> Success {
+        
+        let success = Success()
+        let tabTrayController = self.tabTrayController ?? TabTrayController(tabManager: self.tabManager, profile: self.profile, tabTrayDelegate: self)
+        tabTrayController.changePrivacyMode(isInPrivateMode)
+        
+        tabTrayController.authorisePrivateMode(self.navigationController).uponQueue(dispatch_get_main_queue()) { result in
+            if let _ = result.successValue {
+                self.applyTheme(isPrivate ? Theme.PrivateMode : Theme.NormalMode)
+                
+                if tabTrayController.privateMode != isPrivate {
+                    tabTrayController.changePrivacyMode(isPrivate)
+                }
+                self.tabTrayController = tabTrayController
+            }
+            success.fill(result)
         }
-        self.tabTrayController = tabTrayController
+        
+        return success
```

Comment:
Nit: Instead of declaring a success variable, you can return directly using `bindQueue`. Also, I'm a big fan of using guard to decrease branching so it might be better to use a guard instead of the let _ if.    
```swift
func switchToPrivacyMode(isPrivate isPrivate: Bool) -> Success {
        let tabTrayController = self.tabTrayController ?? TabTrayController(tabManager: self.tabManager, profile: self.profile, tabTrayDelegate: self)
        tabTrayController.changePrivacyMode(isInPrivateMode)
        return tabTrayController.authorisePrivateMode(self.navigationController).bindQueue(dispatch_get_main_queue()) { result in
            guard result.isSuccess else {
                return succeed()
            }
            self.applyTheme(isPrivate ? Theme.PrivateMode : Theme.NormalMode)
            if tabTrayController.privateMode != isPrivate {
                tabTrayController.changePrivacyMode(isPrivate)
            }

            self.tabTrayController = tabTrayController
            return succeed()
        }
    }
```